azkaban-common/src/main/java/azkaban/utils/StdOutErrRedirect.java

```diff
@@ -52,7 +61,40 @@ public LogStream(final OutputStream out, final Level level) {
 
     // Underlying mechanism to log to log4j - all print methods will use this
     private void write(final String string) {
-      logger.log(this.level, string);
+      // If logs are looping, write message to system's stderr then rebind.
+      if (logIsLooping()) {
```

Comment:
How big is the overhead to do this check on every log?

Reply:
Relatively large: https://stackoverflow.com/questions/2347828/how-expensive-is-thread-getstacktrace

I can't really think of a better way to go about this though. Do you have any ideas? I mean, we could run this check randomly with like a 5% chance. But that seems like it's starting to get ridiculous.

Honestly I don't know if there's a good way to recover from these sort of events short of restarting Azkaban. My dream would be able to hard fail Azkaban safely but it holds too much state in memory for that to be an option.

What do you think? Do you have any other ideas?